### PR TITLE
ipc4: add fw_reg definition

### PR DESCRIPTION
**Pull request by RanderWang**
fw_reg is located in memory windows 0 for host to
query fw status

Signed-off-by: Rander Wang <rander.wang@intel.com>

### CODE REVIEW

File: src/platform/tigerlake/include/platform/lib/dma.h [hunk 1]
```diff
@@ -57,6 +57,10 @@
 #define dma_chan_irq(dma, chan) dma_irq(dma)
 #define dma_chan_irq_name(dma, chan) dma_irq_name(dma)
 
+#define HOST_DMA_OUT_CHANNELS	9
+#define GP_DMA_ENGING_COUNT	2
```
Comment (lgirdwood) [inline src/platform/tigerlake/include/platform/lib/dma.h:61]: We already have a macro for number of DW DMA and HDA DMA engines.

File: src/include/ipc4/fw_reg.h [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	 * Stream end offset (LPIB) reported by mixin module allocated on pipeline attached
+	 * to Host Output Gateway during transition from RUNNING to PAUSED. When data
+	 * is not mixed (right after creation/after reset) value "(uint64_t)-1" is reported. When
+	 * first data is mixed then value "0"is reported.
+	 * In number of bytes.
+	 * */
+	uint64_t stream_end_offset;
+} __attribute__((packed, aligned(8)));
+
+#define PV_MAX_SUPPORTED_CHANNELS 8
```
Comment (lgirdwood) [inline src/include/ipc4/fw_reg.h:77]: needs IPC4_ prefix

File: src/include/ipc4/fw_reg.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+		uint32_t reserved : 4;
+	} bits;
+	struct {
+		uint32_t rsvd1: 16;
+		uint32_t type: 12;
+		uint32_t rsvd2: 4;
+	} platform;
+} __attribute__((packed, aligned(4)));
+
+struct FwRegisters {
```
Comment (lgirdwood) [inline src/include/ipc4/fw_reg.h:120]: code style

File: src/include/ipc4/fw_reg.h [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	uint32_t wclk_u;
+} __attribute__((packed, aligned(4)));
+
+struct ipc4_llp_reading_slot {
+	uint32_t node_id;
+	struct ipc4_llp_reading reading;
+} __attribute__((packed, aligned(4)));
+
+union ipc4_rom_info {
+	uint32_t full;
```
Comment (lgirdwood) [inline src/include/ipc4/fw_reg.h:101]: Do we need the full members in each struct/union ?

**Review by lgirdwood — CHANGES_REQUESTED** [review]

**Comment by RanderWang** [comment]
update PR, thanks!

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@RanderWang can you work on upstreaming more of this branch now. Feel free to create a new branch e.g. topic/ipc4 where you can rebase this code into a sequence of patches for fundamental IPC4 protocol in v1.9

**Comment by RanderWang** [comment]
> @RanderWang can you work on upstreaming more of this branch now. Feel free to create a new branch e.g. topic/ipc4 where you can rebase this code into a sequence of patches for fundamental IPC4 protocol in v1.9

sure

**Pull request merged**